### DISCUSSION TITLE
Why opening two port for when connection

### DISCUSSION THREAD

**Discussion by ezzadaghweb** [socketio/socket.io#4940] [Q&A] [answered]
I have a socket.io server, in which i have two namespaces, and as i see when client connect for the first time to the server, two ports are opened in client side. But when connecting to another namespace on the same connection, it still use the first opened port.

My question is, what is the purpose of the second opened port?
<img width="260" alt="Screenshot 2024-02-02 184007" src="https://github.com/socketio/socket.io/assets/17983814/320f3087-5528-4d7b-a8a9-3b9d073a24f7">

**Comment by darrachequesne** [comment]
I think you are referring to the `remotePort` value, which is not the port your server is listening to: https://nodejs.org/api/net.html#socketremoteport

Could you please check?

Reference: https://en.wikipedia.org/wiki/Transmission_Control_Protocol

Edit: `remotePort` instead of `localPort`

**Reply by ezzadaghweb** [reply]
Thank you @darrachequesne, but i think no. Because the console i share is from socket.request.socket._peername object, which is corresponding to connected client's socket as below.

![socket](https://github.com/socketio/socket.io/assets/17983814/b4a3615a-d98c-4053-8422-5d9dc5b62d98)

**Reply by darrachequesne** [reply]
I don't think it's a good idea to rely on a property that starts with an `_`, as it indicates this property is meant to be private and could be removed in newer Node.js versions.

If you need the IP address of the client, you can use the `socket.handshake.address` property.

```js
io.on("connection", (socket) => {
  console.log(socket.handshake.address);
});
```

Reference: https://socket.io/docs/v4/server-api/#sockethandshake

Note: if your server is behind a proxy, you may need to take a look at the `X-Forwarded-For` header.

See also: https://stackoverflow.com/questions/38423930/how-to-retrieve-client-and-server-ip-address-and-port-number-in-node-js

**Reply by ezzadaghweb** [reply]
Thank you very much @darrachequesne .
My purpose was understanding how does socket.io work under the hood :).
I will not use ip or port number, but only i am curious about why when client connect to server, two port was opened at its side and not one port? :)

**Reply by darrachequesne** [reply, marked as answer]
The port (found in either `socket.request.socket._peername.port` or `socket.request.socket.remotePort`) is the port on the client machine used to establish the TCP connection.

You can check it with `netstat` on Linux:

```js
io.on("connection", (socket) => {
  console.log("remotePort", socket.request.socket.remotePort); // prints "remotePort 42620"
  console.log("localPort", socket.request.socket.localPort); // prints "localPort 3000"
});
```

```
$ netstat -an | grep ":3000"
tcp        0      0 127.0.0.1:42620         127.0.0.1:3000          ESTABLISHED
tcp6       0      0 :::3000                 :::*                    LISTEN     
tcp6       0      0 127.0.0.1:3000          127.0.0.1:42620         ESTABLISHED
```

Note: in your example, the first client has two namespaces ("Giris" and "Mektup") which share the same WebSocket connection, which explains why the `remotePort` is the same (45110).

See also: https://stackoverflow.com/questions/3329641/how-do-multiple-clients-connect-simultaneously-to-one-port-say-80-on-a-server

**Reply by ezzadaghweb** [reply]
thank you @darrachequesne , i dont get exactly my answer but you make me on the way :), thank you.
